rust/osm2lanes/src/transform/tags_to_lanes/modes/bus.rs

```diff
@@ -59,85 +59,179 @@ pub(in crate::transform::tags_to_lanes) fn bus(
+    use crate::transform::{RoadWarnings, TagsToLanesMsg};
+
     const BUSWAY: TagKey = TagKey::from("busway");
-    if tags.is(BUSWAY, "lane") {
-        road.forward_outside_mut()
-            .ok_or_else(|| RoadError::unsupported_str("no forward lanes for busway"))?
-            .set_bus(locale)?;
-        if !tags.is("oneway", "yes") && !tags.is("oneway:bus", "yes") {
-            road.backward_outside_mut()
-                .ok_or_else(|| RoadError::unsupported_str("no backward lanes for busway"))?
-                .set_bus(locale)?;
-        }
+    const ONEWAY: TagKey = TagKey::from("oneway");
+
+    #[derive(Debug, PartialEq)]
+    pub(in crate::transform::tags_to_lanes) enum Variant {
+        None,
+        Forward,
+        Backward,
+        Both,
     }
-    if tags.is(BUSWAY, "opposite_lane") {
-        road.backward_outside_mut()
-            .ok_or_else(|| RoadError::unsupported_str("no backward lanes for busway"))?
-            .set_bus(locale)?;
+
+    /// Inferred busway scheme for forward lane and backward lane existing
+    #[derive(Debug)]
+    pub(in crate::transform::tags_to_lanes) struct Scheme(Variant);
+
+    enum Lane {
+        None,
+        Lane,
+        Opposite,
     }
-    if tags.is(BUSWAY + "both", "lane") {
-        road.forward_outside_mut()
-            .ok_or_else(|| RoadError::unsupported_str("no forward lanes for busway"))?
-            .set_bus(locale)?;
-        road.backward_outside_mut()
-            .ok_or_else(|| RoadError::unsupported_str("no backward lanes for busway"))?
-            .set_bus(locale)?;
-        if tags.is("oneway", "yes") || tags.is("oneway:bus", "yes") {
-            return Err(TagsToLanesMsg::ambiguous_str("busway:both=lane for oneway roads").into());
+
+    impl Tags {
+        fn get_bus_lane<T>(&self, key: T, warnings: &mut RoadWarnings) -> Lane
+        where
+            T: AsRef<str>,
+            TagKey: From<T>,
+        {
+            match self.get(&key) {
+                None => Lane::None,
+                Some("lane") => Lane::Lane,
+                Some("opposite_lane") => Lane::Opposite,
+                Some(v) => {
+                    warnings.push(TagsToLanesMsg::unsupported_tag(key, v));
+                    Lane::None
+                },
+            }
         }
     }
-    if tags.is(BUSWAY + locale.driving_side.tag(), "lane") {
-        road.forward_outside_mut()
-            .ok_or_else(|| RoadError::unsupported_str("no forward lanes for busway"))?
-            .set_bus(locale)?;
-    }
-    if tags.is(BUSWAY + locale.driving_side.tag(), "opposite_lane") {
-        return Err(TagsToLanesMsg::ambiguous_tag(
-            BUSWAY + locale.driving_side.tag(),
-            "opposite_lane",
-        )
-        .into());
+
+    impl Scheme {
+        #[allow(clippy::unnecessary_wraps)]
+        pub(in crate::transform::tags_to_lanes) fn from_tags(
+            tags: &Tags,
+            locale: &Locale,
+            oneway: Oneway,
```

Comment:
Takes a oneway and then calculates it below?

Reply:
the oneway coming in is for all vehicles, so I need to first check if there is a value specific for busses.

Now that you mention it, the bus-specific preprocessing should maybe happen one abstraction layer further out, but we can do that later.

Reply:
Ah, I didn't read carefully enough. What if we name these `road_oneway` and `bus_oneway`?